alica_ros_proxy/include/clock/AlicaRosTimer.h

```diff
@@ -0,0 +1,179 @@
+#pragma once
+
+#include "engine/IAlicaTimer.h"
+#include "engine/AlicaClock.h"
+
+#include <condition_variable>
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <ros/callback_queue.h>
+#include <ros/callback_queue_interface.h>
+#include <ros/ros.h>
+
+namespace alicaRosTimer
+{
+
+template <class CallbackQ>
+class SyncStopTimerRosImpl : public std::enable_shared_from_this<SyncStopTimerRosImpl<CallbackQ>>
+{
+    using Base = std::enable_shared_from_this<SyncStopTimerRosImpl<CallbackQ>>;
+public:
+    using TimerCb = std::function<void()>;
+
+    SyncStopTimerRosImpl(CallbackQ& cbQ, TimerCb&& userCb, alica::AlicaTime period)
+            : _userCb(std::move(userCb))
+            , _period(toRosDuration(period))
+            , _nh()
+            , _timer()
+            , _stopMutex()
+            , _stopCv()
+            , _userCbInProgress(false)
+            , _stop(false)
+    {
+        _nh.setCallbackQueue(std::addressof(cbQ));
+    }
+
+    void start()
+    {
+        // Create a one shot timer to fire the first callback immediately. One shot timer is needed because ROS timers don't provide a
+        // a way to create a normal timer whose first callback is immediate
+        // Grab a weak ptr to this object (grabbing a shared_ptr will result in a cycle)
+        _timer = _nh.createTimer(ros::Duration(0), [weak_ptr_self = Base::weak_from_this()](auto&&... args) {
+            // Ensure impl object for this timer is not destroyed
+            if (auto self = weak_ptr_self.lock()) {
+                self->timerCb();
+            }
+        }, true, false);
+        // Don't auto start the timer & instead do it here, because the timerCb can be invoked before _timer is assigned above
+        _timer.start();
+    }
+
+    void stop()
+    {
+        // Block if the timerCb is in progress
+        // TODO: use atomics to check if timerCb is in progress & acquire the mutex only if we need to block
```

Comment:
remove this comment since it is misleading. A mutex is required if we want prevent a stopped timer from being started again in the callback, which can prevent the 1 additional call to timerCb